src/DecksService/Controllers/DecksController.cs

```diff
@@ -0,0 +1,128 @@
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [HttpPut("{deckId}")]
+        public async Task<ActionResult<Deck>> PutAsync(string deckId, [FromBody] Deck deck)
+        {
+            if (!Guid.TryParseExact(deckId, "N", out Guid _))
+            {
+                return new BadRequestResult();
+            }
+
+            deckId = deckId.ToLowerInvariant();
+
+            ActionResult<Deck> result = await this.UpsertAsync(deckId, deck);
+            return result;
+        }
+
+        // DELETE api/decks/[deckId]
+        [ProducesResponseType(StatusCodes.Status204NoContent)] // success or the deck didn't exist in the first place
+        [HttpDelete("{deckId}")]
+        public async Task<IActionResult> DeleteAsync(string deckId)
+        {
+            string userId = GetUserId();
+            bool success = await this.deckInventoryProvider.DeleteDeckAsync(userId, deckId);
+            if (!success)
+            {
+                return new UnauthorizedResult();
+            }
+
+            return new NoContentResult();
+        }
+
+        private async Task<ActionResult<Deck>> UpsertAsync(string deckId, Deck deck)
+        {
+            if (deck == null)
+            {
+                return new BadRequestResult();
+            }
+
+            string userId = GetUserId();
+            string authToken = GetAuthToken(this.Request);
+
+            DeckOperationResult<UpsertResult> operationResult = await this.deckInventoryProvider.UpsertDeckAsync(userId, deckId, deck, authToken);
+            if (operationResult.Result == UpsertResult.NotAuthorizedToModifyExisting)
+            {
+                return new UnauthorizedResult();
+            }
+            else if (operationResult.Result == UpsertResult.CreatedNew)
+            {
+                string path = this.Request.Path;
+                if (!path.EndsWith(deckId, StringComparison.OrdinalIgnoreCase))
+                {
+                    path = path.TrimEnd('/');
+                    path += FormattableString.Invariant($"/{deckId}");
+                }
+                return new CreatedResult(path, operationResult.Deck);
+            }
+            else
+            {
+                return operationResult.Deck;
+            }
+        }
+
+        private string GetUserId()
+        {
+            string userId = User.Claims.Where(x => StringComparer.OrdinalIgnoreCase.Equals(x.Type, @"http://schemas.microsoft.com/identity/claims/objectidentifier")).FirstOrDefault()?.Value;
+            return userId;
+        }
+
+        private static string GetAuthToken(HttpRequest httpRequest)
+        {
+            string authToken = httpRequest.Headers["Authorization"];
+            if (authToken.StartsWith("Bearer", StringComparison.OrdinalIgnoreCase))
+            {
+                authToken = authToken.Substring("Bearer".Length).Trim();
+            }
+            return authToken;
+        }
```

Comment:
At some point we probably want to move this auth-related logic to a common library.